// wdio-lambdatest-service/build/util.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getParentSuiteName = getParentSuiteName;
exports.updateSessionById = updateSessionById;
require("source-map-support/register");
const axios = require('axios');
const {
  version,
  appVersion,
  baseUrl,
  baseUrlApp
} = require("./constants.js");
function getParentSuiteName(fullTitle, testSuiteTitle) {
  const fullTitleWords = fullTitle.split(' ');
  const testSuiteTitleWords = testSuiteTitle.split(' ');
  const shortestLength = Math.min(fullTitleWords.length, testSuiteTitleWords.length);
  let c = 0;
  let parentSuiteName = '';
  while (c < shortestLength && fullTitleWords[c] === testSuiteTitleWords[c]) {
    parentSuiteName += fullTitleWords[c++] + ' ';
  }
  return parentSuiteName.trim();
}
async function updateSessionById(sessionId, data, lambdaCredentials) {
  const sessionUrl = lambdaCredentials.isApp ? `${baseUrlApp}${appVersion.latestVersion}/sessions/${sessionId}` : `${baseUrl}${version.latestVersion}/sessions/${sessionId}`;
  let config = {
    method: 'patch',
    maxBodyLength: Infinity,
    url: sessionUrl,
    headers: {
      'accept': 'application/json',
      'Authorization': `Basic ${Buffer.from(lambdaCredentials.username + ':' + lambdaCredentials.accessKey).toString('base64')}`,
      'Content-Type': 'application/json'
    },
    data: data
  };
  try {
    let response = await axios.request(config);
    if (process.env.LT_CLIENT_LOG === "true") {
      console.log(response);
    }
  } catch (error) {
    console.error(error);
  }
}
